From e1976e0624ffce55fee156801a87671956d26e2e Mon Sep 17 00:00:00 2001
From: Srujana Challa <schalla@marvell.com>
Date: Wed, 7 Jul 2021 14:53:19 +0530
Subject: [PATCH 1629/1921] octeontx2-af: fix cleanup for non cgx mapped PF

Observing "Bcast list, failed to disable PF_FUNC" error for IPsec
device as current code is trying to update mcam entries for non
cgx mapped PF also. So, this patch changes the code to skip updating
of mcam entries for non cgx device while de-intializing
the interface.

Fixes: efeb72a0cd70 ("octeontx2-af: Add multicast/promisc packet
replication table")

Signed-off-by: Srujana Challa <schalla@marvell.com>
Change-Id: I6564ed382d654b7b41a088852a3c3641229f1762
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/55980
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c
index 77de3aaeac28..f808a3f3d195 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c
@@ -2775,6 +2775,7 @@ static int nix_update_mce_rule(struct rvu *rvu, u16 pcifunc,
 	struct npc_mcam *mcam = &rvu->hw->mcam;
 	struct rvu_hwinfo *hw = rvu->hw;
 	struct nix_mce_list *mce_list;
+	int pf;
 
 	/* skip multicast pkt replication for AF's VFs & SDP links */
 	if (is_afvf(pcifunc) || is_sdp_pfvf(pcifunc))
@@ -2783,6 +2784,10 @@ static int nix_update_mce_rule(struct rvu *rvu, u16 pcifunc,
 	if (!hw->cap.nix_rx_multicast)
 		return 0;
 
+	pf = rvu_get_pf(pcifunc);
+	if (!is_pf_cgxmapped(rvu, pf))
+		return 0;
+
 	blkaddr = rvu_get_blkaddr(rvu, BLKTYPE_NIX, pcifunc);
 	if (blkaddr < 0)
 		return -EINVAL;
-- 
2.31.1

